pkg/codegen/hcl2/rewrite_convert.go

```diff
@@ -116,5 +118,105 @@ func RewriteConversions(x model.Expression, to model.Type) model.Expression {
 	if to.AssignableFrom(x.Type()) && sameSchemaTypes(to, x.Type()) {
 		return x
 	}
+	// If we can convert a primitive value in place, do so.
+	if value, ok := convertPrimitiveValues(x, to); ok {
+		return value
+	}
 	return NewConvertCall(x, to)
 }
+
+// convertPrimitiveValues returns a new expression if the given expression can be converted to another primitive type
+// (bool, int, number, string) that matches the target type.
+func convertPrimitiveValues(from model.Expression, to model.Type) (model.Expression, bool) {
+	value := cty.NilVal
+	switch resolveUnderlyingType(to) {
+	case model.BoolType:
+		if stringLiteral, ok := extractStringValue(from); ok {
+			if b, err := strconv.ParseBool(stringLiteral); err == nil {
+				value = cty.BoolVal(b)
+			}
+		}
+	case model.IntType:
+		if stringLiteral, ok := extractStringValue(from); ok {
+			if i, err := strconv.ParseInt(stringLiteral, 10, 64); err == nil {
+				value = cty.NumberIntVal(i)
+			}
+		}
+	case model.NumberType:
+		if stringLiteral, ok := extractStringValue(from); ok {
+			if f, err := strconv.ParseFloat(stringLiteral, 64); err == nil {
+				value = cty.NumberFloatVal(f)
+			}
+		}
+	case model.StringType:
+		stringValue := convertLiteralToString(from)
+		if stringValue != "" {
+			value = cty.StringVal(stringValue)
+		}
+	}
```

Comment:
I'd suggest that you use the `AssignableFrom` method for these checks. This will properly pick up more complicated unions such as those generated by the use `oneOf` in a Pulumi schema. This will also allow you to remove the `resolveUnderlyingType` method.

```suggestion
	// This check is necessary due to #4703
	to = model.ResolveOutputs(to)
	if to == model.DynamicType {
		return nil, false
	}
	switch {
	case to.AssignableFrom(model.BoolType):
		if stringLiteral, ok := extractStringValue(from); ok {
			if v, err := convert.Convert(cty.StringVal(stringLiteral), cty.Bool); err == nil {
				value = v
			}
		}
	case to.AssignableFrom(model.IntType):
		if stringLiteral, ok := extractStringValue(from); ok {
			if v, err := convert.Convert(cty.StringVal(stringLiteral), cty.Bool); err == nil {
				value = v
			}
		}
	case to.AssignableFrom(model.IntType), to.AssignableFrom(model.NumberType):
		if stringLiteral, ok := extractStringValue(from); ok {
			if v, err := convert.Convert(cty.StringVal(stringLiteral), cty.Bool); err == nil {
				value = v
			}
		}
	case to.AssignableFrom(model.StringType):
		if stringValue, ok := convertLiteralToString(from); ok {
			value = cty.StringVal(stringValue)
		}
	default:
		return nil, false
	}
```